File: AE/Arrays/move_element_to_end.py
"""
- input: an array of integers and an integer that represents the element in the array that needs to be moved to the end
- output: an array that has all instances of the integer input moved to the end of the array
- notes: mutatle the input array, do not return a new list
- logic: 
  - going to follow an inverse logic approach; instead of moving the element to the end, going to move elements that are not the input integer to the front
  - in order to do that, we need to keep track of the integers that we need to move to the end, because when we come across an integer that isn't the input integer, we are going to swap the positions, 
  - iterate through the input array using idx:
    - if the num != the integer we need to move, then we know we need to swap:
      - array[idx], array[to_move_pointer] = array[to_move_pointer], array[idx]
      - increment pointer by 1
  - return array
"""


def move_element_to_end(array, toMove):
    to_move_pointer = 0

    for idx in range(len(array)):
        num = array[idx]

        if num != toMove:
            array[idx], array[to_move_pointer] = array[to_move_pointer], array[idx]
            to_move_pointer += 1

    return array

# Time Complexity: O(n)
# Space Complexity: O(1)


def move_element_to_end(array, toMove):
    # two pointers: start of array for values that need to be swapped to the end, and end of array for values that need to be swapped to the front
    i = 0
    j = len(array) - 1

    while i < j:
        # move end pointer to a position where the number needs to be moved to the front
        while i < j and array[j] == toMove:
            j -= 1
        # if we come across a number that needs to be moved, we have already located the position of the number we can swap with
        if array[i] == toMove:
            array[i], array[j] = array[j], array[i]

        i += 1
    return array

# Time Complexity: O(n)
# Space Complexity: O(1)

File: AE/Arrays/moveElementToEnd.js
/*
You're given an array of integers and an integer. Write a function that moves all instances of that integer in the array to the end of the array and returns the array.

The function should perform this in place (e.g it should mutate the input array), and doesn't need to maintain the order of the other integers

Sample Input:
array = [2,1,2,2,2,3,4,2]
toMove = 2

Sample Output:
array = [1,3,4,2,2,2,2,2]

Input: array
Output: array

edge cases: empty array

Brute Force: 
  - Scan array for toMove value, and if found, scan the remaining array for a non-toMove value 
  - swap the two values, and restart initial scan
Time Complexity: O(N^2);
Space Complexity: O(N);

Optimal Solution: 
  - Two pointers: first pointer at index 0 and second pointer at array.length - 1
  - Scan using first pointer in search for toMove value
  - If found, scan using second pointer for a non-toMove value
  - If a non-toMove value is found, swap the two
Time Complexity: O(N);
Space Complexity: O(N);

*/


const moveElementToEndNaive = (array, toMove) => {
  for (let i = 0; i < array.length - 1; i++ ) {
    const value = array[i];

    if (value === toMove) {
      for (let j = i + 1; j < array.length; j++) {
        const potentialSwap = array[j];

        if (potentialSwap !== toMove) {
          [array[i], array[j]] = [array[j], array[i]];
          break
        }
      }
    }
  }
  return array;
}

const moveElementToEndOptimized = (array, toMove) => {
  let left = 0;
  let right = array.length - 1;

  while (left < right) {
    if (array[left] === toMove) {
      while (left < right && array[right] === toMove) right -= 1;

      [array[left], array[right]] = [array[right], array[left]];
    }
    left += 1;
  }
  return array;
}

// Time Complexity: O(N); even though there are two nested loops, the array is being traversed such that each element is explored just once
// Space Complexity: O(1);